first-read-last-write-cache/src/cache.rs

```diff
@@ -130,24 +130,37 @@ impl CacheLog {
     ///     k1 => v1.merge(v1') <- preserves the first read and the last write for 'k1'
     ///     k2 => v2
     ///     k3 => v3
-    pub fn merge(self, mut rhs: Self) -> Result<Self, MergeError> {
-        let mut new_cache = CacheLog::with_capacity(self.log.len() + rhs.log.len());
-
-        for (self_key, self_access) in self.log {
-            match rhs.log.remove(&self_key) {
-                Some(rhs_access) => {
+    pub fn merge(&mut self, rhs: &mut Self) -> Result<(), MergeError> {
+        for (rhs_key, rhs_access) in rhs.log.drain() {
+            match self.log.remove(&rhs_key) {
+                // `Access::merge` should modify the access value in place. This way we will be able to change the log map directly.
+                // Now we are forced to remove the `access` from the map, and insert the new `access` back.
+                // TODO: https://github.com/Sovereign-Labs/sovereign/issues/112
+                Some(self_access) => {
                     let merged = self_access.merge(rhs_access)?;
-                    new_cache.log.insert(self_key, merged);
+                    self.log.insert(rhs_key, merged);
                 }
                 None => {
-                    new_cache.log.insert(self_key, self_access);
+                    self.log.insert(rhs_key, rhs_access);
                 }
-            }
+            };
         }
+        Ok(())
+    }
+
+    /// Clears the cache, returning all key-value pairs as an iterator.
+    pub fn drain(&mut self) -> impl Iterator<Item = (CacheKey, Option<CacheValue>)> + '_ {
+        self.log
+            .drain()
+            .filter_map(|(key, access)| filter_writes(key, access))
```

Comment:
Is it okay to rename it to `drain_all_writes` and update the doc? If we make `filter_map` call site then we would need to make `Access` public, but I  prefer keeping it crate private (so no one will depend on it outside of  the cache crate)

Reply:
Hmm, I would expect `drain_all_writes` to drain the writes but not remove the reads. If we need to keep `Access` private, we could use a NewType wrapper and add methods `is_write() -> bool` and/or `get_written_value() -> Option<CacheValue>`. 

But I'm not sure that we need to do that. It seems like the Access pattern should be part of the crate's API, since that's exactly what it is designed to track. Are you worried that the `Access` enum will change in the future?

Reply:
Talked offline. To minimize the risk of Semver break if we change enum variants later, we're going to try to keep `Access` private for now. We'll rename the method to `clear_cache_and_get_all_writes` or something